question-servers/freeform.js

```diff
@@ -624,10 +640,62 @@ module.exports = {
             data.options.client_files_question_path = path.join(context.question_dir, 'clientFilesQuestion');
 
-            module.exports.processQuestion('render', pc, data, context, (err, courseIssues, _data, html, _fileData, renderedElementNames) => {
-                if (ERR(err, callback)) return;
-                callback(null, courseIssues, html, renderedElementNames);
-            });
+            // This function will render the panel and then cache the results
+            // if cacheKey is not null
+            const doRender = (cacheKey) => {
+                module.exports.processQuestion('render', pc, data, context, (err, courseIssues, _data, html, _fileData, renderedElementNames) => {
+                    if (ERR(err, callback)) return;
+                    if (cacheKey) {
+                        cache.set(cacheKey, {
+                            courseIssues,
+                            html,
+                            renderedElementNames,
+                        });
+                    }
+                    callback(null, courseIssues, html, renderedElementNames, false /* cache miss */);
+                });
+            };
+
+            // This function will check the cache for the specified cache key
+            // and either return the cached render for a cache hit, or render
+            // the panel for a cache miss
+            const getFromCacheOrRender = (cacheKey) => {
+                cache.get(cacheKey, (err, cachedData) => {
+                    // We don't actually want to fail if the cache has an error; we'll
+                    // just render the panel as normal
+                    if (ERR(err, callback)) logger.error(err);
+                    if (!err && cachedData !== null) {
+                        const {
+                            courseIssues,
+                            html,
+                            renderedElementNames,
+                        } = cachedData;
+
+                        callback(null, courseIssues, html, renderedElementNames, true /* cache hit */);
+                    } else {
+                        doRender(cacheKey);
+                    }
+                });
+            };
+
+            if (locals.devMode) {
+                // In dev mode, we should skip caching so that we'll immediately
+                // pick up new changes from disk
+                doRender(null);
+            } else {
+                module.exports._getCacheKey(course, data, (err, cacheKey) => {
+                    // If for some reason we failed to get a cache key, don't
+                    // actually fail the request, just skip the cache entirely
+                    // and render as usual
+                    ERR(err, e => logger.error(e));
+                    logger.info(`Using cache key ${cacheKey}`);
+                    if (err || !cacheKey) {
+                        doRender(null);
+                    } else {
+                        getFromCacheOrRender(cacheKey);
```

Comment:
The control-flow here certainly works, but I'd be slightly more comfortable with explicit callbacks so that we can unwind the stack if there is an error. That is, have `doRender(null, (err) => {...})` and call `callback` explicitly from inside there.